### ISSUE TITLE
[Feature]: virtual module support

### ISSUE THREAD

**Issue by hardfist** [web-infra-dev/rspack#7478]
### What problem does this feature solve?

Virtual module is widely used in bundler ecosystem( vite、webpack), and different bundlers take different approaches,
webpack uses https://www.npmjs.com/package/webpack-virtual-modules?activeTab=readme to support virtual module and vite use https://vitejs.dev/guide/api-plugin.html#virtual-modules-convention to support virtual module
Currently rspack implement a naive implementation of webpack-virtual-modules(https://github.com/rspack-contrib/rspack-plugins/tree/main/packages/plugin-virtual-module) which has some limitation

Since Rspack already supports unplugin so it can actually support both implementation, but too many flexibilities may but ecosystem in chaos(just like vite's virtual module path can be `virtual:my-module` and `\0xxx` which makes it harder to unify plugin usage).

### resolve hook vs input_filesystem
webpack use patched input_filesystem to implement virtual modules which limits virtual module path to be real path url(which may be not bad), so the internal resolver can handle it correctly, vite & rollup use resolve and load and virtual-module-convention to handle virtual path(which is more flexible).

It seems webpack-virtual-modules using patched input_filesystem strategy not a strong limitation in real world usage and is much simpler and performance friendly.

There're 3 ways Rspack can implement virtual module
* Implement builtin rust version of virtualModulePlugin: performance friendly
* Compatible with webpack-virtual-modules, which needs Rspack implement input_filesystem options, which may cause performance regression since every fs call needs trigger napi communication
* Maintain the status quo

### What does the proposed API of configuration look like?

same as https://www.npmjs.com/package/webpack-virtual-modules

**Comment by ahabhgk** [comment]
This probably can split into two layer: memory input filesystem and virtual modules plugin, memory input filesystem for more general use cases and virtual modules plugin is just a wrapper plugin based on memory input filesystem

**Comment by hardfist** [comment]
> This probably can split into two layer: memory input filesystem and virtual modules plugin, memory input filesystem for more general use cases and virtual modules plugin is just a wrapper plugin based on memory input filesystem
yeah, that's the plan, but when you say "memory input filesystem" you mean 
* passing memory input filesystem from js side to rust side and every read will need call js read call
* passing memory input(like object) to rust side and every read will only need rust read call

**Comment by ahabhgk** [comment]
Definitely the second one, but we need to consider to make a better API for user. A MemoryFileSystemRspackPlugin or an enum/options of compiler.inputFileSystem that passes to rust

**Comment by hardfist** [comment]
> Definitely the second one, but we need to consider to make a better API for user. A MemoryFileSystemRspackPlugin or an enum/options of compiler.inputFileSystem that passes to rust

yes, I will make a RFC about InputFilesystem and VirtualModule support

**Comment by hardfist** [comment]
bump

**Comment by watsonhaw5566** [comment]
这个 feature 还有在进行吗， hardfist 兄

**Comment by hardfist** [comment]
> 这个 feature 还有在进行吗， hardfist 兄

no, maybe we will use  inputFs to implement when we solve the performance problem

**Comment by watsonhaw5566** [comment]
bump

**Comment by hardfist** [comment]
closed in #11021
